agent.go

```diff
@@ -0,0 +1,461 @@
+	proc, _, err := p.getRunningProcess(cid, pid)
+	if err != nil {
+		return nil, err
+	}
+
+	var file *os.File
+	if proc.termMaster != nil {
+		file = proc.termMaster
+	} else {
+		if stdout {
+			file = proc.stdout
+		} else {
+			file = proc.stderr
+		}
+	}
+
+	buf := make([]byte, length)
+
+	if _, err := file.Read(buf); err != nil {
+		return nil, err
+	}
+
+	return buf, nil
+}
+
+func (p *pod) initLogger() error {
+	agentLog.Logger.Formatter = &logrus.JSONFormatter{TimestampFormat: time.RFC3339Nano}
+
+	config := newConfig(defaultLogLevel)
+	if err := config.getConfig(kernelCmdlineFile); err != nil {
+		agentLog.WithError(err).Warn("Failed to get config from kernel cmdline")
+	}
+	config.applyConfig()
+
+	agentLog.WithField("version", Version).Info()
+
+	return nil
+}
+
+func (p *pod) initChannel() error {
+	// Check for vsock support.
+	vSockSupported, err := isAFVSockSupported()
+	if err != nil {
+		return err
+	}
+
+	if vSockSupported {
+		// Check if vsock socket exists. We want to cover the case
+		// where the guest OS can support vsock, but the runtime is
+		// still using virtio serial connection.
+		exist, err := vSockPathExist()
+		if err != nil {
+			return err
+		}
+
+		if exist {
+			// Fill the pod with vsock channel information.
+			p.channel = &vSockChannel{}
+
+			return nil
+		}
+	}
+
+	// Open serial channel.
+	file, err := openSerialChannel()
+	if err != nil {
+		return err
+	}
+
+	// Fill the pod with serial channel information.
+	p.channel = &serialChannel{
+		serialConn: file,
+	}
+
+	return nil
+}
+
+func (p *pod) startGRPC() (err error) {
+	var l net.Listener
+
+	if p.isVSock() {
+		l, err = vsock.Listen(1024)
+		if err != nil {
+			return err
+		}
+	} else {
+		socketDir := filepath.Dir(gRPCSockPath)
+		if err = os.MkdirAll(socketDir, fileMode0750); err != nil {
+			return fmt.Errorf("Couldn't create socket directory: %v", err)
+		}
+
+		l, err = net.Listen(gRPCSockScheme, gRPCSockPath)
+		if err != nil {
+			return err
+		}
+	}
+	p.grpcListener = l
+
+	grpcImpl := &agentGRPC{
+		pod: p,
+	}
+
+	grpcServer := grpc.NewServer()
+	pb.RegisterHyperstartServiceServer(grpcServer, grpcImpl)
+
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+
+		grpcServer.Serve(l)
+	}()
+
+	return nil
+}
+
+func (p *pod) loopYamux(session *yamux.Session, grpcConn net.Conn) {
+	defer p.wg.Done()
+
+	for {
+		stream, err := session.Accept()
+		if err != nil {
+			agentLog.Error(err)
+			break
+		}
+		defer stream.Close()
+
+		go func() {
+			io.Copy(grpcConn, stream)
+		}()
+
+		go func() {
+			io.Copy(stream, grpcConn)
+		}()
+	}
+}
+
+func (p *pod) startYamux() error {
+	// If vsock is used, no need to go through yamux.
+	if p.isVSock() {
+		return nil
+	}
+
+	serialChan, ok := p.channel.(*serialChannel)
+	if !ok {
+		return fmt.Errorf("Could not cast into serialChannel type")
+	}
+
+	// Connect gRPC socket.
+	conn, err := net.Dial(gRPCSockScheme, gRPCSockPath)
```

Comment:
Here we dial into the gRPC socket. We know this socket is being listened since `startGRPC()` first listen onto this socket before to give the listener to `server.Serve()` inside the go routine. There is no reason the Dial would fail in that case.

Reply:
yup, you are right.